nslsii/areadetector/xspress3.py

```diff
@@ -0,0 +1,841 @@
+            # orderly manner.
+            # If we are here this is a sign that we have not configured the xs3
+            # correctly and it is expecting to capture more points than it
+            # was triggered to take.
+            while self.capture.get() == 1:
+                i += 1
+                if (i % 50) == 0:
+                    logger.warning("Still capturing data .... waiting.")
+                time.sleep(0.1)
+                if i > 150:
+                    logger.warning("Still capturing data .... giving up.")
+                    logger.warning(
+                        "Check that the xspress3 is configured to take the right "
+                        "number of frames "
+                        f"(it is trying to take {self.parent.cam.num_images.get()})"
+                    )
+                    self.capture.put(0)
+                    break
+
+        except KeyboardInterrupt:
+            self.capture.put(0)
+            logger.warning("Still capturing data .... interrupted.")
+
+        return super().unstage()
+
+    def generate_datum(self, key, timestamp, datum_kwargs):
+        """Create datum kwargs for one Xspress3 channel following a trigger().
+
+        Parameters
+        ----------
+        key: str
+            Xspress3 channel 'name', for example "det1_channels_channel01"
+        timestamp: float
+        datum_kwargs: dict
+
+        Returns
+        -------
+        no return value
+        """
+
+        logger.debug("generate_datum() called with key '%s'", key)
+        # find the channel corresponding to `key`
+        # and create the corresponding datum_kwargs
+        # if we do not find a corresponding channel then we have a problem
+        for channel in self.parent.iterate_channels():
+            if channel.name == key:
+                datum_kwargs.update(
+                    {
+                        "frame": self.parent._abs_trigger_count,
+                        "channel": channel.channel_number,
+                    }
+                )
+                self.mds_keys[channel.channel_number] = key
+                super().generate_datum(
+                    key=key, timestamp=timestamp, datum_kwargs=datum_kwargs
+                )
+                # we are done
+                return
+            else:
+                pass
+
+        # we have a problem
+        # the `key` parameter did not match any of our channels
+        raise ValueError(
+            f"failed to find channel with name '{key}' "
+            f"on Xspress3 detector with PV prefix '{self.parent.prefix}'"
+        )
+
+    # JL: is there any reason to keep this?
+    def configure(self, total_points=0, master=None, external_trig=False, **kwargs):
+        raise NotImplementedError()
+
+    def describe(self):
+        spec_desc = {
+            "external": "FILESTORE:",
+            "dtype": "array",
+            # is there a better value than this for shape?
+            "shape": (self.width.get(),),
```

Comment:
This `shape` is probably a hold-over from the original code. As the comment indicates someone thinks there could be a better shape (might even have been me). I will bring up your suggestion with @tacaswell.

Reply:
This should be what ever the shape of the data that comes back from 1 call to the handler so I would expect this to be either `(4096,)` because below we set this up to be a key-per-channel (if this would be better as 1 key with multiple channels is a different discussion).